cs/src/core/ClientSession/FASTERAsync.cs

```diff
@@ -0,0 +1,365 @@
+                                _ctx = ctx;
+                            }
+
+                            if (!_ctx.markers[EpochPhaseIdx.InProgress])
+                            {
+                                AtomicSwitch(ctx, ctx.prevCtx, _ctx.version);
+                                InitContext(ctx, ctx.prevCtx.guid);
+
+                                // Has to be prevCtx, not ctx
+                                ctx.prevCtx.markers[EpochPhaseIdx.InProgress] = true;
+                            }
+
+                            // Has to be prevCtx, not ctx
+                            if (epoch.MarkAndCheckIsComplete(EpochPhaseIdx.InProgress, ctx.prevCtx.version))
+                            {
+                                GlobalMoveToNextCheckpointState(currentState);
+                            }
+                            break;
+                        }
+                    case Phase.WAIT_PENDING:
+                        {
+                            if (!ctx.prevCtx.markers[EpochPhaseIdx.WaitPending])
+                            {
+                                var notify = (ctx.prevCtx.ioPendingRequests.Count == 0);
+                                notify = notify && (ctx.prevCtx.retryRequests.Count == 0);
+
+                                if (notify)
+                                {
+                                    if (epoch.MarkAndCheckIsComplete(EpochPhaseIdx.WaitPending, ctx.version))
+                                    {
+                                        GlobalMoveToNextCheckpointState(currentState);
+                                    }
+                                    ctx.prevCtx.markers[EpochPhaseIdx.WaitPending] = true;
+                                }
+                            }
+                            else
+                            {
+                                if (epoch.MarkAndCheckIsComplete(EpochPhaseIdx.WaitPending, ctx.version))
+                                {
+                                    GlobalMoveToNextCheckpointState(currentState);
+                                }
+                            }
+                            break;
+                        }
+                    case Phase.WAIT_FLUSH:
+                        {
+                            if (!ctx.prevCtx.markers[EpochPhaseIdx.WaitFlush])
+                            {
+                                bool notify;
+
+                                if (FoldOverSnapshot)
+                                {
+                                    notify = (hlog.FlushedUntilAddress >= _hybridLogCheckpoint.info.finalLogicalAddress);
+                                }
+                                else
+                                {
+                                    notify = (_hybridLogCheckpoint.flushedSemaphore != null) && _hybridLogCheckpoint.flushedSemaphore.CurrentCount > 0;
+                                }
+
+                                if (async && !notify)
```

Comment:
Can you write a comment on why exactly you are doing within `async && !notify` ?